README
fbkeyboard
This is a framebuffer softkeyboard for use on devices with touchscreen input like smartphones.
It was designed for the linux text console.
It uses lib freetype to render fonts.
Keystrokes will be send to the kernel using uinput

How to build:
just run make in this directory

How to run:
./fbkeyboard [font]

usefull tips:
use stty to adjust the console size to avoid overlapping the console and the kayboard.
stty rows <number-of-rows>



debian/changelog
fbkeyboard (0.22) unstable; urgency=medium
  * fix shift key
 -- Uvos <carl@uvos.xyz>  Mon, 14 Jun 2021 10:00:00 +0100

fbkeyboard (0.21) unstable; urgency=medium

  * Use poll() instead of litteraly polling
  * add runlevel and initscript for fbkeyboard shell mode

 -- uvos <devnull@uvos.xyz>  Sat, 03 May 2021 22:14:44 +0100

fbkeyboard (0.2) unstable; urgency=medium

  * Initial Leste packaging

 -- Ivan J. <parazyd@dyne.org>  Sat, 01 May 2021 16:32:44 +0200


draw.h
/*
This file is part of fbkeyboard

Copyright 2017 Julian Winkler <julia.winkler1@web.de>
Copyright 2021 Carl Philipp Klemm <carl@uvos.xyz>

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#pragma once
#include <stddef.h>
#include <stdint.h>
#include <stdbool.h>

bool init_font(const char *font, unsigned int hight);

void fill_rect(char *buf, unsigned int x, unsigned int y, unsigned int xend, unsigned int yend, unsigned int color);

void draw_char(char *buf, unsigned int x, unsigned int y, unsigned int hight, char c);

void draw_text(char *buf, unsigned int x, unsigned int y, unsigned int hight, const char *text);

void draw_keyboard(char *buf, unsigned int width, unsigned int height, int pressed_key, size_t layoutuse);

int init_fb(uint32_t *x_res, uint32_t *y_res, size_t *llength);


layout.c
/*
This file is part of fbkeyboard

Copyright 2017 Julian Winkler <julia.winkler1@web.de>
Copyright 2021 Carl Philipp Klemm <carl@uvos.xyz>

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include "layout.h"

const char *layout[] = { "qwertyuiopasdfghjklzxcvbnm/.",
	"QWERTYUIOPASDFGHJKLZXCVBNM?>",
	"1234567890-={};\'\\,./      /.",
	"!@#$%^&*()_+[]:\"|<>?      ?>"
};

const char *special[SPECIALS] = { "ctl", "esc", "tab", "up", "down", "left", "right" };

const uint16_t keys[ROWS][COLS] =
    { {KEY_Q, KEY_W, KEY_E, KEY_R, KEY_T, KEY_Y, KEY_U, KEY_I, KEY_O, KEY_P, KEY_A, KEY_S, KEY_D, KEY_F, KEY_G, KEY_H,
       KEY_J, KEY_K, KEY_L, KEY_Z, KEY_X, KEY_C, KEY_V, KEY_B, KEY_N, KEY_M, KEY_SLASH, KEY_DOT},
{KEY_1, KEY_2, KEY_3, KEY_4, KEY_5, KEY_6, KEY_7, KEY_8, KEY_9, KEY_0, KEY_MINUS, KEY_EQUAL, KEY_LEFTBRACE,
 KEY_RIGHTBRACE, KEY_SEMICOLON, KEY_APOSTROPHE, KEY_BACKSLASH, KEY_COMMA, KEY_DOT, KEY_SLASH, KEY_X, KEY_C, KEY_V,
 KEY_B, KEY_N, KEY_M, KEY_SLASH, KEY_DOT},
{KEY_LEFTSHIFT, KEY_LEFTCTRL, KEY_ESC, KEY_TAB, KEY_UP, KEY_DOWN, KEY_LEFT, KEY_RIGHT, KEY_SPACE, KEY_ENTER,
 KEY_BACKSPACE}
};


layout.h
/*
This file is part of fbkeyboard

Copyright 2017 Julian Winkler <julia.winkler1@web.de>
Copyright 2021 Carl Philipp Klemm <carl@uvos.xyz>

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#pragma once
#include <stdint.h>
#include <linux/input.h>

#define  ROWS 3
#define  COLS 29
#define  SPECIALS 7

extern const char *layout[];

extern  const char *special[SPECIALS];

extern const uint16_t keys[ROWS][COLS];


uinput.c
/*
This file is part of fbkeyboard

Copyright 2017 Julian Winkler <julia.winkler1@web.de>
Copyright 2021 Carl Philipp Klemm <carl@uvos.xyz>

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include "uinput.h"
#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <fcntl.h>
#include <stdbool.h>

#include "layout.h"

static int fduinput;

bool init_uinput(void)
{
	fduinput = open("/dev/uinput", O_WRONLY);
	if (fduinput == -1) {
		perror("error: cannot open uinput device /dev/uinput");
		return false;
	}
	if (ioctl(fduinput, UI_SET_EVBIT, EV_KEY) == -1) {
		perror("error: SET_EVBIT EV_KEY");
		return false;
	}
	if (ioctl(fduinput, UI_SET_EVBIT, EV_SYN) == -1) {
		perror("error: SET_EVBIT EV_SYN");
		return false;
	}
	for (int key = 0; key < 28; key++)
		ioctl(fduinput, UI_SET_KEYBIT, keys[0][key]);
	for (int key = 0; key < 18; key++)
		ioctl(fduinput, UI_SET_KEYBIT, keys[1][key]);
	for (int key = 0; key < 11; key++)
		ioctl(fduinput, UI_SET_KEYBIT, keys[2][key]);
	struct uinput_user_dev uidev;
	memset(&uidev, 0, sizeof(uidev));
	snprintf(uidev.name, UINPUT_MAX_NAME_SIZE, "fbkeyboard");
	uidev.id.bustype = BUS_USB;
	uidev.id.vendor = 1;
	uidev.id.product = 1;
	uidev.id.version = 1;
	if (write(fduinput, &uidev, sizeof(uidev)) != sizeof(uidev)) {
		fprintf(stderr, "error setting up uinput\n");
		return false;
	}
	if (ioctl(fduinput, UI_DEV_CREATE) == -1) {
		perror("error creating uinput dev");
		return false;
	}
	return true;
}

void uinput_send_modifier(__u16 code, __u16 value)
{
	struct input_event ie = {};
	ie.type = EV_KEY;
	ie.code = code;
	ie.value = value;
	if (write(fduinput, &ie, sizeof(ie)) != sizeof(ie))
		fprintf(stderr, "error sending uinput event\n");
}

void uinput_send_key(__u16 code)
{
	struct input_event ie = {};
	ie.type = EV_KEY;
	ie.code = code;
	ie.value = 1;
	if (write(fduinput, &ie, sizeof(ie)) != sizeof(ie))
		fprintf(stderr, "error: sending uinput event\n");
	ie.value = 0;
	if (write(fduinput, &ie, sizeof(ie)) != sizeof(ie))
		fprintf(stderr, "error: sending uinput event\n");
	ie.code = KEY_LEFTCTRL;
	if (write(fduinput, &ie, sizeof(ie)) != sizeof(ie))
		fprintf(stderr, "error: sending uinput event\n");
	ie.type = EV_SYN;
	ie.code = SYN_REPORT;
	if (write(fduinput, &ie, sizeof(ie)) != sizeof(ie))
		fprintf(stderr, "error: sending uinput event\n");
}


uinput.h
/*
This file is part of fbkeyboard

Copyright 2017 Julian Winkler <julia.winkler1@web.de>
Copyright 2021 Carl Philipp Klemm <carl@uvos.xyz>

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#pragma once
#include <stdbool.h>
#include <linux/uinput.h>

bool init_uinput(void);

void uinput_send_key(__u16 code);

void uinput_send_modifier(__u16 code, __u16 value);


fbkeyboard.c
/*
fbkeyboard.c : framebackbufferfer softkeyboard for touchscreen devices

Copyright 2017 Julian Winkler <julia.winkler1@web.de>
Copyright 2021 Carl Philipp Klemm <carl@uvos.xyz>

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include <stdlib.h>
#include <fcntl.h>
#include <string.h>
#include <unistd.h>
#include <dirent.h>
#include <linux/fb.h>
#include <linux/input.h>
#include <stdint.h>
#include <stdbool.h>
#include <stdio.h>
#include <libinput.h>
#include <errno.h>
#include <libudev.h>
#include <sys/poll.h>

#include "layout.h"
#include "draw.h"
#include "uinput.h"


const char *font = "/usr/share/fonts/truetype/dejavu/DejaVuSans.ttf";

static int open_restricted(const char *path, int flags, void *user_data)
{
	int fd = open(path, flags);
	return fd < 0 ? -errno : fd;
}
 
static void close_restricted(int fd, void *user_data)
{
	close(fd);
}
 
const static struct libinput_interface interface = {
	.open_restricted = open_restricted,
	.close_restricted = close_restricted,
};

int main(int argc, char *argv[])
{
	if (argc != 2) {
		printf("usage: %s [FILE]\nA font file is required.\n", argv[0]);
		return -1;
	}
	
	int layoutuse = 0;

	uint32_t xres, yres;
	size_t line_length;
	int fbfd = init_fb(&xres, &yres, &line_length);
	int hight = (yres / 3 + xres / 2) / 2;

	if (!init_font(argv[1], hight/5))
		return -1;
	
	if (!init_uinput())
		return -2;

	struct udev *udev_cntxt = udev_new();
	struct libinput *li;
	struct libinput_event *event = NULL;
	if (!udev_cntxt)
		return -3;
	
	if(!(li = libinput_udev_create_context(&interface, NULL, udev_cntxt)))
		return -4;
	
	libinput_udev_assign_seat(li, "seat0");
	int lifd = libinput_get_fd(li);
	if (lifd < 0)
		return -5;

	char *backbuffer = malloc(line_length * hight);
	if (backbuffer == 0) {
		perror("OOM!");
		return -1;
	}
	//fill background gray
	fill_rect(backbuffer, 0, 0, xres, hight, 0x444444);
	
	int pressed = -1;
	
	draw_keyboard(backbuffer, xres, hight, pressed, layoutuse);
	lseek(fbfd, line_length * (yres - hight), SEEK_SET);
	write(fbfd, backbuffer, line_length * hight);

	while (true) {
		
		struct pollfd fds;
		fds.fd = lifd;
		fds.events = POLLIN;
	        fds.revents = 0;
		poll(&fds, 1, -1);
		libinput_dispatch(li);
		while((event = libinput_get_event(li)))
		{
			int absolute_x = 0;
			int absolute_y = 0;
			bool touchdown = false;
			bool released = false;
			if (libinput_event_get_type(event) == LIBINPUT_EVENT_TOUCH_DOWN) {
				touchdown = true;
				struct libinput_event_touch *touch = libinput_event_get_touch_event(event);
				if (touch) {
					absolute_x = libinput_event_touch_get_x_transformed(touch, xres);
					absolute_y = libinput_event_touch_get_y_transformed(touch, yres);
				}
			 }
			 else if(libinput_event_get_type(event) == LIBINPUT_EVENT_TOUCH_UP) {
				 released = true;
			 }
			if (released && pressed != -1) {
				if (pressed <= 27)	//normal keys (abc...)
					uinput_send_key(keys[layoutuse >> 1][pressed]);
				else if (pressed == 28) {
					layoutuse ^= 1;
					uinput_send_modifier(KEY_LEFTSHIFT,  layoutuse & 1);
				}
				else if (pressed == 29)
					uinput_send_modifier(KEY_LEFTCTRL, 1);
				else if (pressed <= 38)	//special keys (space, tab, ...)
					uinput_send_key(keys[2][pressed - 28]);
				else if (pressed == 39)	//second page
					layoutuse ^= 2;
				
				pressed = -1;
			}

			
			if (touchdown) {
				switch (((absolute_y - yres) * yres +
					 yres * hight) / (hight/5) / yres) {
				case 0:
					pressed = 29 + absolute_x * 7 / xres;
					break;
				case 1:
					pressed = absolute_x * 10 / xres;
					break;
				case 2:
					if (absolute_x > xres / 20 && absolute_x < xres * 19 / 20)
						pressed = 10 + (absolute_x * 10 - xres / 2) / xres;
					break;
				case 3:
					if (absolute_x < xres * 3 / 20)
						pressed = 28;	//shift
					else if (absolute_x < xres * 17 / 20)
						pressed = 19 + (absolute_x * 10 - xres * 3 / 2) / xres;
					else
						pressed = 38;	//bksp
					break;
				case 4:
					if (absolute_x < xres * 3 / 20)
						pressed = 39;	//123
					else if (absolute_x < xres * 5 / 20)
						pressed = 26;
					else if (absolute_x < xres * 15 / 20)
						pressed = 36;	//space
					else if (absolute_x < xres * 17 / 20)
						pressed = 27;
					else
						pressed = 37;	//enter
					break;
				}
			}
		}
		draw_keyboard(backbuffer, xres, hight, pressed, layoutuse);
		lseek(fbfd, line_length * (yres - hight), SEEK_SET);
		write(fbfd, backbuffer, line_length * hight);
	}
	
	libinput_unref(li);
	udev_unref(udev_cntxt);
	
	return 0;
}


draw.c
/*
This file is part of fbkeyboard

Copyright 2017 Julian Winkler <julia.winkler1@web.de>
Copyright 2021 Carl Philipp Klemm <carl@uvos.xyz>

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include "draw.h"

#include <stdbool.h>
#include <stdio.h>
#include <linux/fb.h>
#include <ft2build.h>
#include <unistd.h>
#include <fcntl.h>
#include FT_FREETYPE_H

#include "layout.h"

static FT_Face face = NULL;

static size_t line_length;

int init_fb(uint32_t *x_res, uint32_t *y_res, size_t *llength)
{
	struct fb_var_screeninfo vinfo;
	struct fb_fix_screeninfo finfo;
	int fbfd = open("/dev/fb0", O_RDWR);
	if (fbfd == -1)
		perror("error: opening framebackbufferfer device /dev/fb0");
	
	if (ioctl(fbfd, FBIOGET_FSCREENINFO, &finfo) == -1) {
		perror("error: reading fixed framebackbufferfer information");
		return -1;
	}
	if (ioctl(fbfd, FBIOGET_VSCREENINFO, &vinfo) == -1) {
		perror("error: reading variable framebackbufferfer information");
		return -1;
	}
	line_length = finfo.line_length;
	
	if (llength)
		*llength = line_length;
	
	if (x_res)
		*x_res = vinfo.xres;
	if (y_res)
		*y_res = vinfo.yres;
	return fbfd;
}


bool init_font(const char *font, unsigned int hight)
{
	static FT_Library library;
	if (FT_Init_FreeType(&library)) {
		perror("error: freetype initialization");
		return false;
	}
	if (FT_New_Face(library, font, 0, &face)) {
		perror("unable to load font file");
		return false;
	}
	if (FT_Set_Pixel_Sizes(face, hight/3, hight/3)) {
		perror("FT_Set_Pixel_Sizes failed");
		return false;
	}
	return true;
}


void fill_rect(char *buf, unsigned int x, unsigned int y, unsigned int xend, unsigned int yend, unsigned int color) {
	int i,j;
	int32_t* line;
	for (i=y; i<yend; i++) {
		line = (int32_t *) (buf + line_length*(i));
		for (j=x; j<xend; j++) {
		*(line + j) = color;
		}
	}
}

void draw_char(char *buf, unsigned int x, unsigned int y, unsigned int hight, char c) {
	unsigned int i, j;
	FT_Load_Char(face, c, FT_LOAD_RENDER);
	y += hight*3/4-face->glyph->bitmap_top;
	for (i=0; i<face->glyph->bitmap.rows; i++)
		for (j=0; j<face->glyph->bitmap.width; j++) {
			*(buf + line_length*(i+y) + (j+x)*4) = *(face->glyph->bitmap.buffer + face->glyph->bitmap.pitch*i + j);
			*(buf + line_length*(i+y) + (j+x)*4+1) = *(face->glyph->bitmap.buffer + face->glyph->bitmap.pitch*i + j);
			*(buf + line_length*(i+y) + (j+x)*4+2) = *(face->glyph->bitmap.buffer + face->glyph->bitmap.pitch*i + j);
		}
}

void draw_text(char *buf, unsigned int x, unsigned int y, unsigned int hight, const char *text) {
	while (*text) {
		draw_char(buf, x, y, hight, *text);
		text++;
		x += face->glyph->advance.x >> 6;
	}
}

void draw_keyboard(char *buf, unsigned int width, unsigned int height, int pressed_key, size_t layoutuse)
{
	int keyWidth = width/10;
	int keyHight = height/5;
	int textHight = keyHight*3/4;
	int textOffset = keyWidth/3;
	int specialKeyWidth = width/7;

	for (size_t key=0; key<7; key++) {
		fill_rect(buf, key*specialKeyWidth, 0, (key+1)*specialKeyWidth-1, keyHight*1-1, (key+29==pressed_key)?0xffffff:0x000000);
		draw_text(buf, key*specialKeyWidth+textOffset, 0, textHight, special[key]);
	}
	
	for (size_t key=0; key<10; key++) {
		fill_rect(buf, key*keyWidth, keyHight*1, (key+1)*keyWidth-1, keyHight*2-1, (key==pressed_key)?0xffffff:0x000000);
		draw_char(buf, key*keyWidth+2+textOffset, keyHight*1, textHight, layout[layoutuse][key]);
	}
	
	for (size_t key=0; key<9; key++) {
		fill_rect(buf, keyWidth/2+key*keyWidth, keyHight*2, keyWidth/2+(key+1)*keyWidth-1, keyHight*3-1, (key+10==pressed_key)?0xffffff:0x000000);
		draw_char(buf, keyWidth/2+key*keyWidth+textOffset, keyHight*2, textHight, layout[layoutuse][key+10]);
	}
	
	fill_rect(buf, 0, keyHight*3, width*3/20-1, keyHight*4-1, (28==pressed_key)?0xffffff:0x000000);
	draw_text(buf, textOffset, keyHight*3, textHight, "shift");

	for (size_t key=0; key<7; key++) {
		fill_rect(buf, width*3/20+key*keyWidth, keyHight*3, width*3/20+(key+1)*keyWidth-1, keyHight*4-1, (key+19==pressed_key)?0xffffff:0x000000);
		draw_char(buf, width*3/20+key*keyWidth+textOffset, keyHight*3, textHight, layout[layoutuse][key+19]);
	}
	
	fill_rect(buf, width*17/20, keyHight*3, width-1, keyHight*4-1, (38==pressed_key)?0xffffff:0x000000);
	draw_text(buf, width*17/20+2+textOffset, keyHight*3, textHight, "bksp");
	fill_rect(buf, 0, keyHight*4, width*3/20-1, keyHight*5-1, (39==pressed_key)?0xffffff:0x000000);
	draw_text(buf, textOffset, keyHight*4, textHight, "?123");
	fill_rect(buf, width*3/20, keyHight*4, width*5/20-1, keyHight*5-1, (26==pressed_key)?0xffffff:0x000000);
	draw_char(buf, width*3/20+2+textOffset, keyHight*4, textHight, layout[layoutuse][26]);
	fill_rect(buf, width/4, keyHight*4, width*3/4-1, keyHight*5-1, (36==pressed_key)?0xffffff:0x000000);
	fill_rect(buf, width*3/4, keyHight*4, width*17/20-1, keyHight*5-1, (27==pressed_key)?0xffffff:0x000000);
	draw_char(buf, width*3/4+2+textOffset, keyHight*4, textHight, layout[layoutuse][27]);
	fill_rect(buf, width*17/20, keyHight*4, width-1, keyHight*5-1, (37==pressed_key)?0xffffff:0x000000);
	draw_text(buf, width*17/20+2+textOffset, keyHight*4, textHight, "enter");
}
